### ISSUE TITLE
How to convert YOLOv8 to a Core ML model with the model type set to “neural network”?

### ISSUE THREAD

**Issue by Chuangvfan** [ultralytics/ultralytics#17889]
### Search before asking

- [X] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

- # The training code I used is as follows

***
from ultralytics import YOLO

model = YOLO('yolov8n.pt')
model.train(data='coco8.yaml', epochs=100, imgsz=640, device='cpu', workers=0, batch=4, cache=True)
***

- # The code I used to convert my trained YOLOv8 model:

***
from ultralytics  import YOLO

torch_model = YOLO('/Users/zhangwufan/Desktop/best.pt')
torch_model.export(format='coreml',nms=True)
***

- # The generated model：

![image](https://github.com/user-attachments/assets/5eb3bf17-e6e3-42a2-981d-ccb3fff010c6)

I noticed that some people get a Model type of **ML Program** when converting, but very few achieve a Model type of **neural network**. This is exactly my concern, as I really want to convert the Model type to **neural network**.
The code from the coremltools official documentation seems incompatible with YOLO models. https://apple.github.io/coremltools/docs-guides/source/convert-pytorch-workflow.html
Here is the link to my reference.

- # coremltools code I tried

![bdad08009ea311893d99cad1307edfe](https://github.com/user-attachments/assets/b88e1b39-b20e-453a-8506-2f665610f1e1)

Could you please let me know what went wrong or how to achieve this? Thank you very much!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Chuangvfan, thank you for your interest in Ultralytics 🚀! We recommend checking out our [Documentation](https://docs.ultralytics.com) for guidance, where numerous examples of [Python usage](https://docs.ultralytics.com/usage/python/) and [model exportation](https://docs.ultralytics.com/modes/export/) are available and may address your query.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us debug. Ensure the training and export process you shared reproduces the observed behavior.

If this is a ❓ Question about exporting YOLOv8 models to Core ML with a "neural network" type, please note that this compatibility depends on the integration between PyTorch and `coremltools`. If the official Core ML tools fail to achieve the desired model type or functionality, we recommend referring to both the [export documentation](https://docs.ultralytics.com/modes/export/) and the [Core ML tools official guide](https://apple.github.io/coremltools/docs-guides/source/convert-pytorch-workflow.html). Additionally, double-check if proper export configurations, supported operations, or framework-specific limitations are influencing your results.

## Upgrade

Please make sure you are using the latest version of `ultralytics` and all dependencies to ensure you have the latest updates and fixes. Update your environment via:

```bash
pip install -U ultralytics
```

## Environments

YOLO can operate in the following up-to-date verified environments (with dependencies preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud VM**: See [Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **AWS Deep Learning AMI**: See [Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**: See [Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Community Support

Engage with the Ultralytics community for further insights into solutions and best practices:
- [Discord](https://discord.com/invite/ultralytics) 🎧 for real-time support.
- [Discourse](https://community.ultralytics.com) for in-depth conversations.
- [Reddit](https://reddit.com/r/Ultralytics) for broader discussions.

## Status Check

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule) tests are passing, verifying correct operation across macOS, Windows, and Ubuntu for supported YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/).

💡 This is an automated response, but rest assured, an Ultralytics engineer will assist you further here soon. Thank you for your patience! 😊

**Comment by Y-T-G** [comment]
What if you don't use `nms=True`?

**Comment by Chuangvfan** [comment]
> 作

If I don’t use` nms=true`, only Non-Maximum Suppression will be removed, but the model type will still be ML Program.

**Comment by Chuangvfan** [comment]
You’re right, and I really appreciate it. However, I noticed that due to compatibility issues, the package has not been adapted to the latest version of NumPy. Using numpy==1.24.4 is a correct option.

> 啊，我认为你需要使用`format="mlmodel"`

**Comment by Y-T-G** [comment]
`mlmodel` is a legacy format. `mlpackage` is what coremltools recommends.

https://stackoverflow.com/questions/78229310/difference-between-mlmodel-and-mlpackage

**Comment by Chuangvfan** [comment]
thank you again

> `mlmodel` is a legacy format. `mlpackage` is what coremltools recommends.
> 
> https://stackoverflow.com/questions/78229310/difference-between-mlmodel-and-mlpackage

**Comment by glenn-jocher** [comment]
You're welcome! Indeed, `mlpackage` is the recommended Core ML format for newer models due to its support for advanced functionalities, as highlighted by Apple and coremltools. Let us know if you encounter any issues.

**Comment by rromanchuk** [comment]
@Y-T-G that hasn't been true for > at least two years.  

Literally, just try it.  `yolo export model=yolo11n.pt format=coreml nms=True`


```swift
let mlModel = try await yolo11n.load(configuration: MLModelConfiguration()).model
let model = try VNCoreMLModel(for: mlModel)
let request = VNCoreMLRequest(model: model) { request, error in
                if let recognizedObjects = (request.results as? [VNRecognizedObjectObservation]) {
                    debugPrint(recognizedObjects) // <-- actually look at the results
                }
            }
let handler = VNImageRequestHandler(cgImage: /** <your simple cg image> **/)
try await handler.perform([request])
```

Or just open it in Xcode and drag and drop an image into the preview. 

I've only ever seen nms flag work with Neural networks  (and coreml tools 6) `yolo export model=yolo11n.pt format=mlmodel nms=True`

> You're welcome! Indeed, mlpackage is the recommended Core ML format for newer models due to its support for advanced functionalities, as highlighted by Apple and coremltools. Let us know if you encounter any issues.

 LLM, mlpackage is a container, which can package both neural networks and ML Programs. Simply double click your .mlmodel and XCode will prompt to convert to an mlpackage, not to be confused with converting a neural network to ML Program.

**Comment by Y-T-G** [comment]
@rromanchuk I am not sure which reply you're responding to here.
